File: src-tauri/src/directory_nav_service/dtos/sort_files_by_dto.rs
use serde::{Deserialize, Serialize};
use super::super::models::date_range::DateRange;

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")] 
pub enum OldestNewest {
    Oldest,
    Newest,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")] 
pub enum LargestSmallest {
    Largest,
    Smallest,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "PascalCase")]
pub struct SortFilesByDTO{
    pub date_modified_range: Option<DateRange>,
    pub date_modified: Option<OldestNewest>,

    pub date_created_range: Option<DateRange>,
    pub date_created: Option<OldestNewest>,

    pub extensions: Vec<String>,

    pub size: Option<LargestSmallest>,

    /** If `false`, then the results will only include directories. If `None`, then nothing happens */
    pub files_only: Option<bool>
}

File: src/app/core/dtos/sort-files-by-dto.ts
import { DateRange } from "@core/models/date-range";

type OldestNewest = "oldest" | "newest";
type LargestSmallest = "largest" | "smallest";
export interface SortFilesByDTO{
    DateModifiedRange:DateRange | undefined,
    DateModified: OldestNewest | undefined,

    DateCreatedRange:DateRange | undefined,
    DateCreated: OldestNewest | undefined,

    Extensions: string[],

    Size: LargestSmallest | undefined, 
    /** If `false`, then no directories will be included in the results */
    FilesOnly: boolean | undefined
}

/** Returns a new interface will all undefined values */
export function defaultSortFilesByDTO():SortFilesByDTO{
    return {
        DateModifiedRange: undefined,
        DateModified: undefined,

        DateCreatedRange: undefined,
        DateCreated: undefined,

        Extensions: [],

        Size: undefined,

        FilesOnly: undefined
    }
}